gitfile/tokenfetcher_spi.go

```diff
@@ -0,0 +1,181 @@
+
+	"github.com/mshaposhnik/service-provider-integration-scm-file-retriever/gitfile/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SpiTokenFetcher token fetcher implementation that looks for token in the specific ENV variable.
+type SpiTokenFetcher struct {
+	k8sClient client.Client
+	namespace string
+}
+
+const (
+	namespaceFile = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	letterBytes   = "abcdefghijklmnopqrstuvwxyz1234567890"
+)
+
+func NewSpiTokenFetcher() *SpiTokenFetcher {
+	namespace, err := ioutil.ReadFile(namespaceFile)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	scheme := runtime.NewScheme()
+	if err = corev1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	if err = v1beta1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	// creates the client
+	k8sClient, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		panic(err.Error())
+	}
+	return &SpiTokenFetcher{k8sClient: k8sClient, namespace: string(namespace)}
+}
+
+func (s *SpiTokenFetcher) BuildHeader(ctx context.Context, repoUrl string, loginCallback func(url string)) (*HeaderStruct, error) {
+
+	var tBindingName = "file-retriever-binging-" + randStringBytes(6)
+	var secretName = "file-retriever-secret-" + randStringBytes(5)
+
+	// create binding
+	newBinding := newSPIATB(tBindingName, s, repoUrl, secretName)
+	err := s.k8sClient.Create(ctx, newBinding)
+	if err != nil {
+		zap.L().Error("Error creating Token Binding item:", zap.Error(err))
+		return nil, err
+	}
+
+	// scheduling the binding cleanup
+	defer func() {
+		// clean up token binding
+		err = s.k8sClient.Delete(ctx, newBinding)
+		if err != nil {
+		}
+	}()
+
+	// now re-reading SPITokenBinding to get updated fields
+	var tokenName string
+	for {
+		readBinding := &v1beta1.SPIAccessTokenBinding{}
+		err = s.k8sClient.Get(ctx, client.ObjectKey{Namespace: s.namespace, Name: tBindingName}, readBinding)
+		if err != nil {
+			zap.L().Error("Error reading TB item:", zap.Error(err))
+			return nil, err
+		}
+		tokenName = readBinding.Status.LinkedAccessTokenName
+		if tokenName != "" {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("task is cancelled")
+		default:
+			time.Sleep(200 * time.Millisecond)
+		}
+	}
+	zap.L().Info(fmt.Sprintf("Access Token to watch: %s", tokenName))
+
+	// now try read SPIAccessToken to get link
+	var url string
+	var loginCalled = false
+	for {
+		readToken := &v1beta1.SPIAccessToken{}
+		_ = s.k8sClient.Get(ctx, client.ObjectKey{Namespace: s.namespace, Name: tokenName}, readToken)
+		if readToken.Status.Phase == v1beta1.SPIAccessTokenPhaseAwaitingTokenData && !loginCalled {
+			url = readToken.Status.OAuthUrl
+			zap.L().Info(fmt.Sprintf("URL to OAUth: %s", url))
+			go loginCallback(url)
```

Comment:
Why do you call the login asynchronously? We are waiting for the result anyway, right?

Reply:
Also, the callback should use the same context so that the caller can cancel the whole thing...